### PR TITLE
grpc: report connectivity state changes on the ClientConn for Subscribers

### PR DESCRIPTION
**Pull request by my4-dev**
This PR adds a logic to report connectivity state changes on the ClientConn for Subscribers.

Please refer to https://github.com/grpc/grpc-go/pull/6036#issuecomment-1488895547 in https://github.com/grpc/grpc-go/pull/6036 . Original issue is https://github.com/grpc/grpc-go/issues/5818.

RELEASE NOTES: none

### CODE REVIEW

File: internal/internal.go [hunk 1]
```diff
@@ -122,6 +122,9 @@ var (
 	// deleted or changed.
 	BinaryLogger interface{} // func(binarylog.Logger) grpc.ServerOption
 
+	// AddConnectivityStateSubscriber adds a grpcsync.Subscriber to a provided grpc.ClientConn
+	AddConnectivityStateSubscriber interface{} // func(*grpc.ClientConn, grpcsync.Subscriber)
+
```
Comment (easwars) [inline internal/internal.go:127]: s/AddConnectivityStateSubscriber/SubscribeToConnectivityStateChanges/

File: clientconn_test.go [hunk 2]
```diff
@@ -1079,6 +1080,113 @@ func (s) TestDefaultServiceConfig(t *testing.T) {
...
+		// side blocks until settings are received.
+		time.Sleep(100 * time.Millisecond)
+		framer := http2.NewFramer(conn, conn)
+		close(sent)
+		if err := framer.WriteSettings(http2.Setting{}); err != nil {
+			t.Errorf("Error while writing settings. Err: %v", err)
+			return
+		}
+		<-dialDone // Close conn only after dial returns.
+	}()
```
Comment (easwars) [inline clientconn_test.go:1130]: Please write this as an end-to-end style test in the `grpc/test` package. 
- Create a new file `connectivity_state_updates_test.go`
- Create a test backend service. See: https://github.com/grpc/grpc-go/blob/master/internal/stubserver/stubserver.go#L219
- Create a manual resolver
- Create a ClientConn that dials the above test service and uses the manual resolver
- Subscribe to connectivity state updates
- Ensure the initial state is as expected
- Push the address of the test backend through the manual resolver
- Ensure that you can make an RPC to the test backend
- Ensure that all expected states upto READY are received by the subscriber
- Close the clientConn
- Ensure that SHUTDOWN is received by the subscriber.


See https://github.com/grpc/grpc-go/blob/master/test/idleness_test.go to get an idea of how to create the test backend, how to create the manual resolver, how to push updates using the manual resolver etc.

Please let me know if you have any questions here. Thanks.

**Review by easwars — CHANGES_REQUESTED** [review]

**Comment by my4-dev** [comment]
@easwars : Thank you for reviewing! I've fixed some implementations according to your comments.

File: clientconn.go [hunk 3]
```diff
@@ -137,7 +138,9 @@ func (dcs *defaultConfigSelector) SelectConfig(rpcInfo iresolver.RPCInfo) (*ires
 func DialContext(ctx context.Context, target string, opts ...DialOption) (conn *ClientConn, err error) {
 	cc := &ClientConn{
 		target: target,
-		csMgr:  &connectivityStateManager{},
+		csMgr: &connectivityStateManager{
+			pubSub: grpcsync.NewPubSub(),
+		},
```
Comment (easwars) [inline clientconn.go:143]: Sorry, I probably should have been more explicit. What I meant was to have a function to create the `connectivityStateManager`.

```golang
func newConnectivityStateManager(id *channelz.Identifier) *connectivityStateManager {
  return &connectivityStateManager{
    channelzID: id,
    pubSub: grpcsync.NewPubSub(),
  },
}
```

And in this function, instead of initializing the `csMgr` field here, wait until the call to `cc.channelzRegistration(target)` down below.

Also, change `cc.channelzRegistration()` to not set `cc.csMgr.channelzID = cc.channelzID`.

File: clientconn.go [hunk 4]
```diff
@@ -561,6 +569,9 @@ func (csm *connectivityStateManager) updateState(state connectivity.State) {
 		return
 	}
 	csm.state = state
+	if csm.pubSub != nil {
```
Comment (easwars) [inline clientconn.go:572]: Since the `pubSub` field is initialized at creation time, we dont need this `nil` check here.

File: test/connectivity_state_updates_test.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	t.Cleanup(func() { cc.Close() })
+
+	// Start a test backend and push an address update via the resolver.
+	backend := stubserver.StartTestService(t, nil)
+	t.Cleanup(backend.Stop)
+
+	s := newTestSubscriber()
+	internal.SubscribeToConnectivityStateChanges.(func(cc *grpc.ClientConn, s grpcsync.Subscriber) func())(cc, s)
+
+	expectedStates := []connectivity.State{
```
Comment (easwars) [inline test/connectivity_state_updates_test.go:74]: Nit: s/expectedStates/wantStates/

File: test/connectivity_state_updates_test.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+		connectivity.Idle,
+		connectivity.Shutdown,
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
+		for _, expectedState := range expectedStates {
+			select {
+			case actualState := <-s.onMsgCh:
```
Comment (easwars) [inline test/connectivity_state_updates_test.go:86]: Nit: s/actualState/gotState/

File: test/connectivity_state_updates_test.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+					t.Errorf("Received unexpected state: %q; want: %q", actualState, expectedState)
+				}
+			case <-time.After(defaultTestTimeout):
+				t.Error("Timeout when expecting the onMessage() callback to be invoked")
+			}
+			if t.Failed() {
+				break
+			}
+		}
+		wg.Done()
```
Comment (easwars) [inline test/connectivity_state_updates_test.go:97]: Nit: here and in the below goroutine have `defer wg.Done()` as the first line. That way, there will be no codepath where we will ever forget to call `wg.Done()`.

File: test/connectivity_state_updates_test.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+		cc.Close()
+		awaitState(ctx, t, cc, connectivity.Shutdown)
+
+		wg.Done()
+	}()
+
+	wg.Wait()
+	if t.Failed() {
+		t.FailNow()
+	}
```
Comment (easwars) [inline test/connectivity_state_updates_test.go:120]: Can this be done instead in the check for `t.Failed` in the first goroutine. Because I don't see any call to `t.Error` in the second one. So, it might just be easier to fail from the first goroutine when an unexpected state is received.

File: test/connectivity_state_updates_test.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+func (s) TestConnectivityStateUpdates(t *testing.T) {
+	// Create a ClientConn with a short idle_timeout.
+	r := manual.NewBuilderWithScheme("whatever")
+	dopts := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithResolvers(r),
+		grpc.WithIdleTimeout(defaultTestShortIdleTimeout),
+		grpc.WithDefaultServiceConfig(`{"loadBalancingConfig": [{"round_robin":{}}]}`),
```
Comment (easwars) [inline test/connectivity_state_updates_test.go:59]: We probably don't see a custom service config for this test.

File: test/connectivity_state_updates_test.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+		t.Fatalf("grpc.Dial() failed: %v", err)
+	}
+	t.Cleanup(func() { cc.Close() })
+
+	// Start a test backend and push an address update via the resolver.
+	backend := stubserver.StartTestService(t, nil)
+	t.Cleanup(backend.Stop)
+
+	s := newTestSubscriber()
+	internal.SubscribeToConnectivityStateChanges.(func(cc *grpc.ClientConn, s grpcsync.Subscriber) func())(cc, s)
```
Comment (easwars) [inline test/connectivity_state_updates_test.go:72]: Minor nit: Maybe we should register the subscriber as soon as the ClientConn is created. There is no reason to wait for the backend server to be created before registering this subscriber.

File: test/connectivity_state_updates_test.go (hunk 8 above)
Comment (my4-dev) [reply test/connectivity_state_updates_test.go:120]: If so, the first goroutine will only fail, not the main goroutine.
We need to fail after `wg.Wait()` in order to wait finishing the first goroutine.

File: clientconn.go [hunk 11]
```diff
@@ -525,8 +526,8 @@ func chainStreamClientInterceptors(cc *ClientConn) {
 		chainedInt = func(ctx context.Context, desc *StreamDesc, cc *ClientConn, method string, streamer Streamer, opts ...CallOption) (ClientStream, error) {
 			return interceptors[0](ctx, desc, cc, method, getChainStreamer(interceptors, 0, streamer), opts...)
 		}
+		cc.dopts.streamInt = chainedInt
```
Comment (easwars) [inline clientconn.go:529]: This seems like an unrelated change. Why is this required?

File: test/connectivity_state_updates_test.go [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+	cc, err := grpc.Dial(r.Scheme()+":///test.server", dopts...)
+	if err != nil {
+		t.Fatalf("grpc.Dial() failed: %v", err)
+	}
+	t.Cleanup(func() { cc.Close() })
+
+	s := newTestSubscriber()
+	internal.SubscribeToConnectivityStateChanges.(func(cc *grpc.ClientConn, s grpcsync.Subscriber) func())(cc, s)
+
+	// Start a test backend and push an address update via the resolver.
```
Comment (easwars) [inline test/connectivity_state_updates_test.go:69]: The address update is only pushed in the goroutine down below. So, we can skip that in this comment.

File: test/connectivity_state_updates_test.go [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+		connectivity.Ready,
+		connectivity.Idle,
+		connectivity.Shutdown,
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		for _, expectedState := range wantStates {
```
Comment (easwars) [inline test/connectivity_state_updates_test.go:84]: s/expectedState/wantState

File: clientconn.go (hunk 11 above)
Comment (my4-dev) [reply clientconn.go]: I'm sorry, this is an accident.
I've reverted this change.

File: clientconn.go [hunk 14]
```diff
@@ -584,6 +601,12 @@ func (csm *connectivityStateManager) getNotifyChan() <-chan struct{} {
 	return csm.notifyChan
 }
 
+func (csm *connectivityStateManager) close() {
+	csm.mu.Lock()
+	defer csm.mu.Unlock()
```
Comment (easwars) [inline clientconn.go:606]: We don't need to hold the lock here since the `pubSub` field is initialized at creation time and is never written to after that.

File: test/connectivity_state_updates_test.go [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+		awaitState(ctx, t, cc, connectivity.Idle)
+
+		cc.Close()
+		awaitState(ctx, t, cc, connectivity.Shutdown)
+	}()
+
+	wg.Wait()
+	if t.Failed() {
+		t.FailNow()
+	}
```
Comment (easwars) [inline test/connectivity_state_updates_test.go:117]:
```suggestion
	doneCh := make(chan struct{})
	go func() {
		defer close(doneCh)
		for _, wantState := range wantStates {
			select {
			case gotState := <-s.onMsgCh:
				if gotState != wantState {
					t.Errorf("Received unexpected state: %q; want: %q", gotState, wantState)
				}
			case <-time.After(defaultTestTimeout):
				t.Error("Timeout when expecting the onMessage() callback to be invoked")
			}
			if t.Failed() {
				break
			}
		}
	}()

	// Verify that the ClientConn moves to READY.
	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
	defer cancel()
	r.UpdateState(resolver.State{Addresses: []resolver.Address{{Addr: backend.Address}}})
	awaitState(ctx, t, cc, connectivity.Ready)

	// Verify that the ClientConn moves to IDLE as there is no activity.
	awaitState(ctx, t, cc, connectivity.Idle)

	cc.Close()
	awaitState(ctx, t, cc, connectivity.Shutdown)
	
	<-doneCh
	if t.Failed() {
		t.FailNow()
	}
```

File: test/connectivity_state_updates_test.go [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+	}()
+
+	go func() {
+		defer wg.Done()
+		r.UpdateState(resolver.State{Addresses: []resolver.Address{{Addr: backend.Address}}})
+
+		// Verify that the ClientConn moves to READY.
+		ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+		defer cancel()
+		awaitState(ctx, t, cc, connectivity.Ready)
```
Comment (easwars) [inline test/connectivity_state_updates_test.go:105]: Since `awaitState` calls `t.Fatal`, I think we have this piece of code execute as part of the main goroutine instead of a separate one. I have made a suggestion for this change. Please see if that works. Thanks.

**Review by easwars — APPROVED** [review]

**Comment by easwars** [comment]
@dfawley for second set of eyes.

File: test/connectivity_state_updates_test.go [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+				break
+			}
+		}
+	}()
+
+	// Verify that the ClientConn moves to READY.
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	r.UpdateState(resolver.State{Addresses: []resolver.Address{{Addr: backend.Address}}})
+	awaitState(ctx, t, cc, connectivity.Ready)
```
Comment (dfawley) [inline test/connectivity_state_updates_test.go:100]: This looks like it will be flaky to me, just like the previous instance of the same thing we found.

However, idleness is a great test of the pubsub delivering transitions reliably that the other API could miss.  So I think we can just delete this `awaitState` and everything will be good.

File: test/connectivity_state_updates_test.go [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+	doneCh := make(chan struct{})
+	go func() {
+		defer close(doneCh)
+		for _, wantState := range wantStates {
+			select {
+			case gotState := <-s.onMsgCh:
+				if gotState != wantState {
+					t.Errorf("Received unexpected state: %q; want: %q", gotState, wantState)
+				}
+			case <-time.After(defaultTestTimeout):
```
Comment (dfawley) [inline test/connectivity_state_updates_test.go:87]: Nit: initialize the context at the start of the test and use `<-ctx.Done()` here instead.

File: test/connectivity_state_updates_test.go [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+	// Verify that the ClientConn moves to IDLE as there is no activity.
+	awaitState(ctx, t, cc, connectivity.Idle)
+
+	cc.Close()
+	awaitState(ctx, t, cc, connectivity.Shutdown)
+
+	<-doneCh
+	if t.Failed() {
+		t.FailNow()
+	}
```
Comment (dfawley) [inline test/connectivity_state_updates_test.go:111]: IIUC this can be deleted since there is no code after it.

**Comment by my4-dev** [comment]
Thanks @dfawley !
I've applied your comments!

**Review by dfawley — APPROVED** [review]
